src/main/middleware/Authentication.ts

```diff
@@ -0,0 +1,56 @@
+import { NextFunction } from 'express';
+import { Request, Response } from 'express';
+import { auth } from 'firebase-admin/lib/auth';
+import Auth = auth.Auth;
+
+/**
+ * Middleware function to check if the request has correct client header using fireabse-admin auth
+ */
+export class FirebaseAuth {
+  private admin: Auth;
+
+  constructor(firebaseAdmin: Auth) {
+    this.admin = firebaseAdmin;
+  }
+
+  isAuthorizedForNewUser = async (
```

Comment:
This function seems pretty similiar to `isAuthorizedForProtection` - is there a reason you separated the two? Is it possible to capture the duplication into one function?

Reply:
This is because I want to avoid setting `uid` to request body except to when making new user. Other endpoints do not have to have or should not contain `uid` in request body, so I made two different authorization functions in order to keep the same functionalities as much as possible.

Reply:
Hmm, would it be possible to set a flag somehow as a parameter to indicate that this is a call with making a new user?

Reply:
I don't think we can do a flag for a parameter because these are middle ware functions, I think they need to have those specific parameters. 

@yunishimura0716 For the `isAuthorizedForNewUser`, won't the request body already contain `idFirebase` when we call `POST /promotions`? So then we don't need to set the `idFirebase` in the request body? https://github.com/ubclaunchpad/PromoPal-backend/blob/973b08e20c38c72acdd771da4f30c73ef8ef15ff/src/main/middleware/Authentication.ts#L27

![image](https://user-images.githubusercontent.com/49849754/109110505-27fa3300-76ec-11eb-822c-41a2a4ce3842.png)

Reply:
@ericrkuo I did not implement for protected endpoint for `promotion`, but yeah, that's right. Because we can get `uid` (idFirebase here) once the `idToken` verified correctly, I want to set the `idFirebase` after success of authorization. However, if you think there's nothing problem to set `uid` at FE to request body, we can simplify the middleware function.

Reply:
Oops sorry, I meant `POST /users`, fixed the comment above. Is it safe to send the firebase id in the request body?

Reply:
I guess yes. The reason why the request body doesn't contain `uid` is because just want to make sure they can get `uid` after the `idToken` is verified. However, the `idToken` is provided from FE whose current user is signed up/in. That means, `uid` in request body should be correct and we do not have to get `uid` in BE.

Reply:
Ahh okay, yes then in that case I think we should be good. Because if we were able to successfully verify the `idToken`, then we can assume the `firebaseId` in the request body is valid. I think we can simplify the middleware to be all the same then, would that be okay?

Reply:
sure! I'll refactor it!

Reply:
Awesome thanks Yu!